Assignement in condition warning

RblSb opened the pull request:
There is also some idea to skip warning when you wrap condition in additional parentheses in first commit.
Related to https://github.com/HaxeFoundation/haxe/issues/8634
```haxe
var foo = false;
var foo2 = false;
// warning for these cases
var value = (foo = foo2) ? true : false;
if (foo = true) {}
if (foo = null) {}
if ({123; foo = true;}) {}

while (foo = true) {}
do {} while (foo = true);

var dyn:Dynamic = false;
if (dyn = null) {}
if (dyn = false) {}

if ((foo = null) && (foo = true)) {}
if ((foo = null) && ((foo = true) && (foo = true))) {}

// no warning
if ({foo = true; true;}) {}
```

RblSb committed:
Assignement in condition warning

RblSb committed:
Assignement in condition warning

kLabz commented:
That double parenthesis syntax seems out of place :/

RblSb committed:
version without parentheses

RblSb committed:
version without parentheses

RblSb commented:
I will write test if you guys think that this is good warning / implementation to have

kLabz commented:
For the `Dynamic` (ie `is_bool` is false) case, I think it makes sense.

The bool case is more problematic. Assignment of a variable of type `T` being itself of type `T` is by design, so writing `if (a = true) { ... }` is totally valid.

Maybe add a sub warning for the bool case, that would be off by default?

RblSb committed:
sub-warning

`bool = null`, `dyn = dyn2` are reported

`bool = false`, `bool = bool2` is opt in

RblSb committed:
sub-warning

`bool = null`, `dyn = dyn2` are reported

`bool = false`, `bool = bool2` is opt in

RblSb commented:
Waiting for a re-review!
